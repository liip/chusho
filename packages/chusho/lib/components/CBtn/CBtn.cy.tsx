import { ChushoUserOptions } from '../../types';

import { CBtn } from '.';

describe('CBtn', () => {
  it('applies local and global classes', () => {
    cy.mount(
      <CBtn class="btn" data-test="btn" variant="primary">
        Label
      </CBtn>,
      {
        global: {
          provide: {
            $chusho: {
              options: {
                components: {
                  btn: {
                    class({ variant }) {
                      return [
                        'config-btn',
                        { 'config-btn-primary': variant?.primary },
                      ];
                    },
                  },
                },
              } as ChushoUserOptions,
            },
          },
        },
      }
    );

    cy.get('[data-test="btn"]').should(
      'have.class',
      'btn config-btn config-btn-primary'
    );

    cy.getWrapper().then((wrapper) => {
      wrapper.setProps({ bare: true });

      cy.get('[data-test="btn"]').should(
        'not.have.class',
        'config-btn config-btn-primary'
      );
    });
  });

  it('applies the right attributes', () => {
    cy.mount(<CBtn data-test="btn">Label</CBtn>);

    cy.get('[data-test="btn"]').should('have.attr', 'type', 'button');

    cy.getWrapper().then((wrapper) => {
      wrapper.setProps({ type: 'submit' });

      cy.get('[data-test="btn"]').should('have.attr', 'type', 'submit');
    });
  });

  it('renders the slot', () => {
    cy.mount(<CBtn data-test="btn">Label</CBtn>);

    cy.get('[data-test="btn"]').should('have.text', 'Label');
  });

  it('should be a link if prop `href` is set', () => {
    cy.mount(
      <CBtn data-test="link" href="https://chusho.dev" type="button" disabled>
        Label
      </CBtn>
    );

    cy.get('a[data-test="link"]')
      .should('have.attr', 'href', 'https://chusho.dev')
      .and('not.have.attr', 'type', 'button')
      .and('not.have.attr', 'disabled', 'disabled');
  });

  it('should be a RouterLink if prop `to` is set', () => {
    cy.mount(
      <CBtn data-test="link" to="/home">
        Label
      </CBtn>,
      {
        global: {
          stubs: {
            RouterLink: true,
          },
        },
      }
    );

    cy.get('router-link-stub[data-test="link"]').should(
      'have.attr',
      'to',
      '/home'
    );
  });

  it('should be a NuxtLink if prop `to` is set in a Nuxt app', () => {
    cy.mount(
      <CBtn data-test="link" to="/home">
        Label
      </CBtn>,
      {
        global: {
          mocks: {
            $nuxt: {},
          },
          stubs: {
            NuxtLink: true,
          },
        },
      }
    );

    cy.get('nuxt-link-stub[data-test="link"]').should(
      'have.attr',
      'to',
      '/home'
    );
  });
});
