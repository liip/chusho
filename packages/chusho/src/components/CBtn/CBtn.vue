<script lang="ts">
import Vue, { VNodeData } from 'vue';
import { mergeData } from 'vue-functional-data-merge';
import { VNode } from 'vue/types/umd';

interface BtnProps {
  href: string;
  type: string;
  variant: string;
  disabled: boolean;
}

export default Vue.extend<BtnProps>({
  name: 'CBtn',

  functional: true,

  props: {
    /**
     * Make the button a link to the given location
     */
    href: {
      type: String,
      default: null,
    },

    /**
     * In case the button is a `button` element, specify its type
     */
    type: {
      type: String,
      default: 'button',
    },

    /**
     * Customize the button appearance by applying one or multiple variants defined in the config (btn.variants).
     * Example: `"primary large"`
     */
    variant: {
      type: String,
      default: '',
    },

    /**
     * Disable the button (doesnâ€™t apply to links) and apply the associated class from the config (btn.disabled).
     */
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  render(h, { props, data, parent, children }): VNode {
    const { btn } = parent.$chusho.options;
    const tag = props.href ? 'a' : 'button';

    const classes = [btn.default, { [btn.disabled]: props.disabled }];
    const variants = props.variant.split(' ');
    variants.forEach(variant => {
      const target = btn.variants[variant];
      if (target) classes.push(target);
    });

    const componentData: VNodeData = {
      class: classes,
      attrs: {
        href: props.href,
        // Concerns only on button tags, skip for anchors
        ...(props.disabled && tag === 'button' && { disabled: true }),
        ...(props.type && tag === 'button' && { type: props.type }),
      },
    };

    return h(tag, mergeData(data, componentData), children);
  },
});
</script>
