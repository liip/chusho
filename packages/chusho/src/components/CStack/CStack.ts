import Vue, { VNode } from 'vue';
import { warn } from '@/utils/debug';

interface StackProps {
  gap: string;
  containerTag: string;
  itemTag: string;
}

export default Vue.extend<StackProps>({
  name: 'CStack',

  functional: true,

  props: {
    /**
     * One of the gaps defined in the config (stack.gaps) to apply to the element.
     */
    gap: {
      type: String,
      required: true,
    },
    /**
     * The tag used for the container, example: `ul`
     */
    containerTag: {
      type: String,
      default: 'div',
    },
    /**
     * The tag used for the items, example: `li`
     */
    itemTag: {
      type: String,
      default: 'div',
    },
  },

  render(h, { children, props, data, parent }): VNode {
    const stackConfig = parent?.$chusho?.options?.components?.stack;
    const gap = stackConfig?.gaps?.[props.gap];

    if (!gap) {
      warn(
        `Cannot find a Stack "gap" named "${props.gap}" in the config, you must define it before referencing it.`
      );
      return h('div', data, children);
    }

    // Wrap each children in a container to apply the gap
    children = children.map(
      (child: VNode): VNode => {
        // Skip empty/whitespace-only text nodes
        return child.tag || (child.text && child.text.trim().length)
          ? h(props.itemTag, { class: gap.itemClass }, [child])
          : child;
      }
    );

    data.class = [data.class, gap.containerClass];

    return h(props.containerTag, data, children);
  },
});
