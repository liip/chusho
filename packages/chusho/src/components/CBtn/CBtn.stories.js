import StoryRouter from 'storybook-vue-router';
import CBtn from './CBtn';

export default {
  title: 'Components|Btn',
  component: CBtn,
  parameters: {
    componentSubtitle:
      'Uniformized button style for `router-link`, `nuxt-link`, `a` or `button` elements.',
    options: {
      componentConfig: [
        {
          name: 'defaultClass',
          type: { summary: 'string' },
          description: 'Class applied to all CBtn components.',
        },
        {
          name: 'disabledClass',
          type: { summary: 'string' },
          description:
            'Class applied to Btn when prop `disabled` is set to `true`.',
        },
        {
          name: 'variants',
          type: { summary: 'object' },
          description:
            'Predefined set of variants.\n\nFor example: `{ "primary": "bg-blue-500 text-white" }`',
        },
      ],
    },
  },
  decorators: [
    StoryRouter(
      {},
      {
        routes: [
          {
            path: '/',
            name: 'home',
            component: { template: `<div>Home</div>` },
          },
        ],
      }
    ),
  ],
};

export const Default = () => ({
  components: { CBtn },
  template: '<CBtn>Click me</CBtn>',
});

export const AsLink = () => ({
  components: { CBtn },
  template: '<CBtn href="#">Click me</CBtn>',
});

export const AsRouterLink = () => ({
  components: { CBtn },
  template: `<CBtn :to="{ name: 'home' }">Click me</CBtn>`,
});

export const TypeSubmit = () => ({
  components: { CBtn },
  template: '<CBtn type="submit">Click me</CBtn>',
});

export const WithVariant = () => ({
  components: { CBtn },
  template: '<CBtn variant="default medium">Click me</CBtn>',
});

export const Disabled = () => ({
  components: { CBtn },
  template: `<CBtn disabled>Click me</CBtn>`,
});

export const WithCustomClasses = () => ({
  components: { CBtn },
  template: `<CBtn class="font-bold">Click me</CBtn>`,
});
