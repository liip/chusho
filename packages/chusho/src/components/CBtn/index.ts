import CBtn from './CBtn.vue';

export { CBtn };
export default CBtn;
