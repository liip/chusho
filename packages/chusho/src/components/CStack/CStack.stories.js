import CStack from './CStack.vue';
import CBtn from '../CBtn';

export default {
  title: 'Stack',
  component: CStack,
  parameters: {
    componentSubtitle:
      'Automatically stack child components with configurable gap between each item.',
    options: {
      componentConfig: 'stack',
    },
  },
};

export const Default = () => ({
  components: { CStack },
  template: `<CStack gap="1">
    <span>One</span>
    Two
    <span>Three</span>
  </CStack>`,
});

export const WithCustomTags = () => ({
  components: { CStack },
  template: `<CStack gap="2" containerTag="ul" itemTag="li">
    <span>One</span>
    Two
    <span>Three</span>
  </CStack>`,
});

export const WithSubComponent = () => ({
  components: { CStack, CBtn },
  template: `<CStack gap="2">
    <span>One</span>
    <CBtn>Two</CBtn>
    <span>Three</span>
  </CStack>`,
});
