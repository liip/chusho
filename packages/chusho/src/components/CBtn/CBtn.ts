import {
  defineComponent,
  h,
  inject,
  mergeProps,
  PropType,
  resolveComponent,
} from 'vue';
import { RouteLocationRaw } from 'vue-router';

import { DollarChusho } from '../../types';
import { generateConfigClass } from '../../utils/components';
import { warn } from '../../utils/debug';
import componentMixin from '../mixin';

export default defineComponent({
  name: 'CBtn',

  mixins: [componentMixin],

  inheritAttrs: false,

  props: {
    /**
     * Make the button a link to the given URL.
     */
    href: {
      type: String,
      default: null,
    },

    /**
     * Use a `router-link` (or `nuxt-link` in a Nuxt app) in the background, works juste like the `router-link` prop of the same name.
     *
     * Note that you can also pass all other `router-link` props such as `activeClass`, `replace`, … See [Vue router documentation](https://router.vuejs.org/api/#router-link) for a complete list of available props.
     */
    to: {
      type: [String, Object] as PropType<RouteLocationRaw>,
      default: null,
    },

    /**
     * Specifies the button type, does not apply when `to` or `href` props are used. Example: `submit`.
     */
    type: {
      type: String,
      default: 'button',
    },

    /**
     * Disable the button, does not apply when `to` or `href` props are used.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  render() {
    const btnConfig = inject<DollarChusho | null>('$chusho', null)?.options
      ?.components?.btn;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let tag: any = 'button';
    let extraAttrs: Record<string, unknown> = {};

    if (this.to) {
      tag = resolveComponent('router-link');

      if (tag === 'router-link') {
        warn(
          `CBtn got a “to” prop but RouterLink component couldn’t be found. Is Vue Router installed? To make a CBtn behave like a traditional link, use the prop “href” instead.`
        );
      }

      extraAttrs = {
        to: this.to,
      };
    } else if (this.href) {
      tag = 'a';
      extraAttrs = {
        href: this.href,
      };
    }

    const elementProps: Record<string, unknown> = {
      ...extraAttrs,
      // Concerns only on button tags, skip for anchors
      ...(this.disabled && tag === 'button' && { disabled: true }),
      ...(this.type && tag === 'button' && { type: this.type }),
      ...generateConfigClass(btnConfig?.class, this.$props),
    };

    return h(tag, mergeProps(this.$attrs, elementProps), this.$slots);
  },
});
