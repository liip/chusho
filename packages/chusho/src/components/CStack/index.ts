import CStack from './CStack.vue';

export { CStack };
export default CStack;
