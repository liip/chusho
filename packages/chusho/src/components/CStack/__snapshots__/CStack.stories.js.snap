// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Stack Default 1`] = `
<div
  class="-mt-1"
>
  <div
    class="mt-1"
  >
    <span>
      One
    </span>
  </div>
  <div
    class="mt-1"
  >
    
    Two
    
  </div>
  <div
    class="mt-1"
  >
    <span>
      Three
    </span>
  </div>
</div>
`;

exports[`Storyshots Stack With Custom Tags 1`] = `
<ul
  class="-mt-2"
>
  <li
    class="mt-2"
  >
    <span>
      One
    </span>
  </li>
  <li
    class="mt-2"
  >
    
    Two
    
  </li>
  <li
    class="mt-2"
  >
    <span>
      Three
    </span>
  </li>
</ul>
`;

exports[`Storyshots Stack With Sub Component 1`] = `
<div
  class="-mt-2"
>
  <div
    class="mt-2"
  >
    <span>
      One
    </span>
  </div>
   
  <div
    class="mt-2"
  >
    <button
      class="inline-block py-2 px-4 bg-green-200 text-green-900 rounded"
      type="button"
    >
      Two
    </button>
  </div>
   
  <div
    class="mt-2"
  >
    <span>
      Three
    </span>
  </div>
</div>
`;
